File: Data Structures & Algorithms/trapping-rain-water/submission-1.py
class Solution:
    def trap(self, height: List[int]) -> int:
        stack = []  # Store indices of bars
        total_water = 0

        for current in range(len(height)):
            # While we have a higher bar that can trap water
            while stack and height[current] > height[stack[-1]]:
                top = stack.pop()

                if not stack:
                    break

                # Calculate trapped water between bars
                distance = current - stack[-1] - 1
                bounded_height = min(height[current], height[stack[-1]]) - height[top]
                total_water += distance * bounded_height

            stack.append(current)

        return total_water

File: Data Structures & Algorithms/daily-temperatures/submission-1.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;
        int[] res = new int[n];
        int[] st = new int[n];
        int top = -1;
        for(int i = 0; i < n;i++){
            int t = temperatures[i];
            while(top >= 0 && t > temperatures[st[top]]){
                int j = st[top--];
                res[j] = i - j;
            }
            st[++top] = i;
        }
        return res;
    }
}
